conference aware

tarekbecker opened the pull request:
FYI

tarekbecker committed:
conference aware

tarekbecker committed:
progress conference aware

tarekbecker committed:
cleanup + conference aware

tarekbecker commented:
Status:
- grundlegend funktioniert das mit der Konferenz
- es können neue angelegt werden, meine angesehen werden, zu welchen beitreten
- Beitritt per Name + Shared Secret (wie moodleeinschreibeschlüssel)
- JEDER der joined ist attendee
- Author, Reviewer, Chair kann vom Chair gesetzt werden
- Wer eine Konferenz erstellt ist chair und attendee
- ein chair kann andere chairs ernennen, aber nicht sich selbst entfernen (damit wir keine Zombie Konferenzen haben)
- ein user kann submissions anlegen (title). Submission ist mit der aktuellen konferenz verknüpft
- konferenz kann gewechselt werden, einfach auf irgenwas von user gehen (was nicht konferenz abhängig ist) und dann im User dropdown auf dne namen der konferenz klicken
- die default konferenz (nahc login) kann gesetzt werden, ACHTUNG button aktualisiert sich erst nach reload, da der callback nicht richtig funktioniert
- alle rechte funktionen (AUTHOR, CHAIR; REVIEWER). Achtung: sind Konferenzabhängig.
- wenn ich meine eigenen rechte änder (e.g. + autor) ist ein neuladen nötig, login/logout ist nicht notwendig
- tags, affiliations sind noch nicht konferenzabhängig (können wir diskutieren, ob wir das wollen)
  - Idee: Nur der Chair kann tags festlegen (hat vor und nachteile)
  - Wenn nicht: wer hat das recht welche zu löschen / verwalten?
- Ordner Struktur nach Rollen: USER (alle user funktionen: profile, conference mgmt), AUTHOR, CHAIR, REVIEWER (sollte klar sein)
- ich habe alle leeren controller gelöscht sowie leere Seiten (e.g. statistics); können wir wieder hinzufügen, wenn es jemand implementiert
- **BITTE einen controller / datei**

@Coksnuss @huze169 imho ist das ein guter Stand zum mergen. 

**INFO ZU PROMISES**: Das ist kein gültiger Promise Code (auch wenn er funktioniert)

``` javascript
AuthService.getUser().then(function (userData) {
  $scope.user = userData;
});

function getSubmissions() {
  // $scope.user.id ist undefined wenn AuthService noch keinen Wert geliefert hat
  User
    .submissions({id: $scope.user.id, filter: { where: {conferenceId: $stateParams.conferenceId}}})
   .$promise
   .then(function(result) {
      $scope.submissions = result;
      $scope.count = result.length;
   });
 }

getSubmission();
```

Code sollte so aussehen

``` javascript
function getSubmissions() {
   AuthService.getUserId().then(function(userid) {
     // promise is ausgewertet, Wert MUSS da sein!
     // sonst würde .catch aufgerufen werden
     User
      .submissions({id: userId, filter: { where: {conferenceId: $stateParams.conferenceId}}})
      .$promise
      .then(function(result) {
        $scope.submissions = result;
        $scope.count = result.length;
       });
    }
});
```

tarekbecker committed:
promises

tarekbecker committed:
fixed login

schnz committed:
Use kebab-case for tags (for compatibility of upcoming pull request with on-the-fly tag creation)

Coksnuss commented:
Sieht schonmal gut aus!

Feedback UI:
* Beitreten einer Konferenz: Ähnlich wie bei Moodle, würde ich hier die verfügbaren Konferenzen auflisten, so dass nur der Beitrittsschlüssel eingegeben werden muss (anstatt zusätzlich noch der Konferenzname)

schnz commented:
Kann nicht alles testen da im initDb command nicht alles angelegt wird und du das anlegen der Nutzerrollen wohl absichtlich auskommentiert hast(?)
Aber hier schonmal das Feedback zu dem was ich bisher so gesehen habe. Weiteres Feedback folgt dann wenn alles läuft. Und sorry für das späte Feedback :/. Hoffe beim nächsten mal gehts schneller.

Feedback (Functional):
- Beitreten einer Konferenz: Ähnlich wie bei Moodle würde ich die verfügbaren Konferenzen über eine Dropdown Liste selektierbar machen, so dass der Benutzer nur noch den Beitrittsschlüssel angeben muss (anstatt zusätzlich noch den Konferenznamen [Tippfehler vorprogrammiert])
- Das initDb script fügt nicht alle benötigten Daten ein. Es werden nur die tags und affiliations angelegt. Es fehlt der chair user, die roles (gewollt, da auskommentiert(?)), conferences, ...

Feedback (Code):
Ich weiß, wir werden nicht den schönsten und tollsten code am Schluss haben. Aber da wir im Team arbeiten würde ich mich schon gerne an ein paar best practice Regeln halten damit jeder mit dem Code auch produktiv arbeiten kann.
- Logik aus den views raushalten und in den Controller stecken (z.B. in https://github.com/tarekauel/tkpraktikum/blob/039d1fc0b298b7890896677fee29d8b3782b02e1/client/views/user/joinConference.html#L6)
- Die services ihrem Zweck entsprechend nutzen und nicht "missbrauchen" (z.B. https://github.com/tarekauel/tkpraktikum/blob/039d1fc0b298b7890896677fee29d8b3782b02e1/client/js/controllers/user/joinConference.js#L8 - Soweit ich sehe wird die zugewiesene Methode nicht im $scope benötigt)

tarekbecker committed:
fixed url name

tarekbecker commented:
Also:
1. user erstellen
2. login
3. create conference 
4. my conference checken, ob conference da ist
5. im chair menü auf den conference namen klicken, das wählt die Konferenz aus
6. unter conference management user management wählen
7. author zuweisen
8. submission anlegen

die rollen gibt es nciht mehr, sind jetzt entitäten.

Chair ist eine Entität die einen User und eine Conference verbindet (geht leider nciht schöner glaub ich)

schnz committed:
Removed unnecessary fields when creating author/reviewer/chair

schnz committed:
Validate uniqueness of conference name.

schnz committed:
Fixed initDatabase scripts

schnz committed:
Merge branch 'tarekauel-conference_awareness' into conference_awareness_enhancements

tarekbecker committed:
don't cache user promise

schnz committed:
Merge branch 'tarekauel-conference_awareness' into conference_awareness_enhancements

tarekbecker commented:
- Beitreten einer Konferenz mit DropDown (fügen wir später hinzu)
- initDb fliegt raus und wir nehmen einfach ein default json das wir kopieren (siehe UI Branch)

Code:
- zu 1. ja müssen wir mal überall sauber refactorn (ist einfach kopiert aus dem Login)
- zu 2. noch geändert

schnz commented:
> initDb  fliegt raus und wir nehmen einfach ein default json das wir kopieren (siehe UI Branch)

Habe ich jetzt schon in einem anderen PR gefixt. Können wir die Paradigmen nicht einfach beibehalten die wir haben?

> zu 2. noch geändert

Ich meine das eher Prinzipiell. Ist mir auch z.B. hier aufgefallen (Stichwort: Namespace pollution)
https://github.com/tarekauel/tkpraktikum/blob/039d1fc0b298b7890896677fee29d8b3782b02e1/client/js/app.js#L156

Ich gehe aber gerade den ganzen PR durch und änder das alles ab und fixe das was beim testen nicht funktioniert.

Commit:
Merge pull request #3 from Coksnuss/conference_awareness_enhancements

Use kebab-case for tags (for compatibility of upcoming pull request)

schnz committed:
initDb script: Added conference attendance relations

schnz committed:
Fixed hasMany relations

schnz committed:
AuthService: Changed back getUser to be a method because the internal user object might change during the application lifecycle! (via login() / logout())

schnz committed:
Refactored AuthService
- Removed exposure of currentConferenceId through $rootScope to keep AuthService self-contained.
- Removed unused injections in AuthService
- Removed internal rolesCache variable. User object now holds the required data
- Drastically reduced required requests to resolve roles inside AuthService
- Removed dependency on window.localStorage

HeaderController:
- Removed $rootScope dependency
- Fixed typo in initialization variable
- Fixed indentation
- Made $scope.changeConference working

ConferenceController:
- Removed $rootScope dependency
- Removed $scope namespace pollution (assigned service was never used elsewhere)
- Removed unecessary $watchCollection() call

schnz committed:
Removed unecessary state permission declarations

Commit:
Merge pull request #5 from Coksnuss/conference_awareness_enhancements

Conference awareness enhancements
